indra/llcommon/lua_function.cpp

```diff
@@ -684,6 +705,38 @@ std::pair<LuaFunction::Registry&, LuaFunction::Lookup&> LuaFunction::getState()
     return { registry, lookup };
 }
 
+/*****************************************************************************
+*   source_path()
+*****************************************************************************/
+lua_function(source_path, "return the source path of the running Lua script")
+{
```

Comment:
What's going to happen shall L be nullptr?

Also I assume `lua_function` is a macro hiding the real function signature. I'm not sure if I like that or not. But at least a better comment why it seems void yet returns 1 would be helpful

Reply:
Coding a new Lua-accessible C++ entry point involves a certain amount of boilerplate. I'm particularly bothered by frameworks that require you to (a) code a function and then (b) register that function at runtime in some other function far removed, and this is one such. It's too easy to code a new function in all sincerity, and then spend debugging time trying to figure out why your Lua script can't use it.

`lua_function.h` [defines a macro](https://github.com/secondlife/viewer/blob/lua-startup/indra/llcommon/lua_function.h#L161-L171) that declares and instantiates [a `LuaFunction` subclass](https://github.com/secondlife/viewer/blob/lua-startup/indra/llcommon/lua_function.h#L133-L142) and introduces the definition of its `static int call(lua_State* L)` method. Initializing the static instance takes care of registering the particular entry point function.

Every Lua entry point accepts exactly `(lua_State*)` and returns `int`, as described in [the Lua manual](https://www.lua.org/manual/5.1/manual.html#lua_CFunction). Incoming parameters are found on the Lua stack via operations on the `lua_State*`, return values are pushed on the same stack. The returned `int` specifies the number of stack entries to be considered return values.

If the Lua VM calls any such entry point with a null `lua_State*`, things are already badly busted. If anyone else intentionally calls one of these entry points, it's a coding error.